### PR TITLE
[Android] Adjust GUI SRD peak luminance when display is in HDR PQ mode

### PR DESCRIPTION
**Pull request by thexai**
## Description
[Android] Adjust GUI SRD peak luminance when display is in HDR PQ mode

This the equivalent functionality of https://github.com/xbmc/xbmc/pull/18984 and https://github.com/xbmc/xbmc/pull/21973 for Android.

## Motivation and context
This is requested a lot on the forums....

https://forum.kodi.tv/showthread.php?tid=376360
https://forum.kodi.tv/showthread.php?tid=357205
https://www.reddit.com/r/kodi/comments/1awb6h3/osd_hdr_too_bright_during_playback/

With this is not necessary set subtitles color dark grey or different color for SDR / HDR and also is adjusted OSD and all GUI elements.

Android already does basic tone mapping - there is no need for full tone mapping from BT.709 to BT.2020 or gamma to PQ transfer. Only is need to set the SDR peak to 100 or 200 nits because by default it seems to be set to 1000 nits (at least in Shiled) and this is why it looks much brighter than normal SDR.

When SDR peak is set to 100 - 200 nits then GUI has same appearance than in SDR and GUI white is regular SDR white.

The default setting is 40% and matches what we already have in Windows. That is, the result is consistent with Windows using the same setting value. Adjusting GUI setting to 100% gives you the maximum brightness (~1000 nits), which is the same as before this PR.

## How has this been tested?
Runtime Shield Pro 2019

Forum users:
- Fire TV Stick 4K Max 2nd gen (https://forum.kodi.tv/showthread.php?tid=376360&pid=3186614#pid3186614)
- Fire TV Stick 4K Max 1st gen (https://forum.kodi.tv/showthread.php?tid=376360&pid=3186768#pid3186768)

## What is the effect on users?
No more tricks to adjust brightness of subtitles / OSD / GUI when playing HDR10 or Dolby Vision contents.

## Screenshots (if appropriate):

screenshot00001

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by sarbes** [comment]
I can't say I'm too happy about this approach. It introduces colour inaccuracies and will result in heavy banding, having just ~6.5 bits per color channel. It might also lengthen the execution time of the shaders, depending on the platform.

Pretty much every SOC with HDR support comes with its dedicated mapping hardware, so that such hacks are not needed.

For devices with lacking drivers, a slightly more reasonable approach would be a proper resolve pass at the end, preferably using pixel local storage or a framebuffer fetch. This would at least avoid the colour inaccuracies.

**Comment by Hitcher** [comment]
Run some tests and it's not working for some Dolby Vision videos.

Working:
```
Video
ID                                       : 1
Format                                   : HEVC
Format/Info                              : High Efficiency Video Coding
Format profile                           : Main 10@L5.1@Main
HDR format                               : Dolby Vision, Version 1.0, Profile 8.1, dvhe.08.06, BL+RPU, HDR10 compatible / SMPTE ST 2086, Version HDR10, HDR10 compatible
Codec ID                                 : V_MPEGH/ISO/HEVC
Duration                                 : 1 h 15 min
Bit rate                                 : 21.3 Mb/s
Width                                    : 3 840 pixels
Height                                   : 1 920 pixels
Display aspect ratio                     : 2.000
Frame rate mode                          : Constant
Frame rate                               : 23.976 (24000/1001) FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:0 (Type 0)
Bit depth                                : 10 bits
Bits/(Pixel*Frame)                       : 0.121
Stream size                              : 11.2 GiB (97%)
Writing library                          : x265 3.5+1-f0c1022b6:[Linux][GCC 10.2.1][64 bit] 10bit
Encoding settings                        : cpuid=1176575 / frame-threads=4 / wpp / no-pmode / no-pme / no-psnr / no-ssim / log-level=2 / input-csp=1 / input-res=3840x1920 / interlace=0 / total-frames=0 / level-idc=0 / high-tier=0 / uhd-bd=0 / ref=3 / no-allow-non-conformance / repeat-headers / annexb / aud / no-eob / no-eos / hrd / info / hash=0 / no-temporal-layers / no-open-gop / min-keyint=96 / keyint=96 / gop-lookahead=0 / bframes=4 / b-adapt=2 / b-pyramid / bframe-bias=0 / rc-lookahead=96 / lookahead-slices=8 / scenecut=40 / no-hist-scenecut / radl=0 / no-splice / no-intra-refresh / ctu=64 / min-cu-size=8 / no-rect / no-amp / max-tu-size=32 / tu-inter-depth=1 / tu-intra-depth=1 / limit-tu=0 / rdoq-level=0 / dynamic-rd=0.00 / no-ssim-rd / signhide / no-tskip / nr-intra=0 / nr-inter=0 / no-constrained-intra / strong-intra-smoothing / max-merge=3 / limit-refs=1 / no-limit-modes / me=1 / subme=2 / merange=57 / temporal-mvp / no-frame-dup / no-hme / weightp / weightb / no-analyze-src-pics / deblock=0:0 / no-sao / no-sao-non-deblock / rd=3 / selective-sao=0 / early-skip / rskip / no-fast-intra / no-tskip-fast / no-cu-lossless / b-intra / no-splitrd-skip / rdpenalty=0 / psy-rd=2.00 / psy-rdoq=0.00 / no-rd-refine / no-lossless / cbqpoffs=0 / crqpoffs=0 / rc=crf / crf=11.0 / qcomp=0.60 / qpstep=4 / stats-write=0 / stats-read=0 / vbv-maxrate=21900 / vbv-bufsize=40000 / vbv-init=0.9 / min-vbv-fullness=50.0 / max-vbv-fullness=80.0 / crf-max=0.0 / crf-min=0.0 / ipratio=1.40 / pbratio=1.30 / aq-mode=2 / aq-strength=1.00 / cutree / zone-count=0 / no-strict-cbr / qg-size=32 / no-rc-grain / qpmax=69 / qpmin=0 / no-const-vbv / sar=1 / overscan=0 / videoformat=5 / range=0 / colorprim=9 / transfer=16 / colormatrix=9 / chromaloc=1 / chromaloc-top=0 / chromaloc-bottom=0 / display-window=0 / master-display=G(13250,34500)B(7500,3000)R(34000,16000)WP(15635,16450)L(10000000,1) / cll=662,211 / min-luma=0 / max-luma=1023 / log2-max-poc-lsb=8 / vui-timing-info / vui-hrd-info / slices=1 / no-opt-qp-pps / no-opt-ref-list-length-pps / no-multi-pass-opt-rps / scenecut-bias=0.05 / hist-threshold=0.03 / no-opt-cu-delta-qp / no-aq-motion / hdr10 / hdr10-opt / no-dhdr10-opt / no-idr-recovery-sei / analysis-reuse-level=0 / analysis-save-reuse-level=0 / analysis-load-reuse-level=0 / scale-factor=0 / refine-intra=0 / refine-inter=0 / refine-mv=1 / refine-ctu-distortion=0 / no-limit-sao / ctu-info=0 / no-lowpass-dct / refine-analysis-type=0 / copy-pic=1 / max-ausize-factor=1.0 / no-dynamic-refine / no-single-sei / no-hevc-aq / no-svt / no-field / qp-adaptation-range=1.00 / scenecut-aware-qp=0conformance-window-offsets / right=0 / bottom=0 / decoder-max-rate=0 / no-vbv-live-multi-pass
Default                                  : Yes
Forced                                   : No
Color range                              : Limited
Color primaries                          : BT.2020
Transfer characteristics                 : PQ
Matrix coefficients                      : BT.2020 non-constant
Mastering display color primaries        : Display P3
Mastering display luminance              : min: 0.0001 cd/m2, max: 1000 cd/m2
Maximum Content Light Level              : 662 cd/m2
Maximum Frame-Average Light Level        : 211 cd/m2
```

Not working:
```
Video
ID                                       : 1
Format                                   : HEVC
Format/Info                              : High Efficiency Video Coding
Format profile                           : Main 10@L5@High
HDR format                               : Dolby Vision, Version 1.0, Profile 5, dvhe.05.06, BL+RPU
Codec ID                                 : V_MPEGH/ISO/HEVC
Duration                                 : 40 min 19 s
Bit rate                                 : 24.0 Mb/s
Width                                    : 3 840 pixels
Height                                   : 1 920 pixels
Display aspect ratio                     : 2.000
Frame rate mode                          : Constant
Frame rate                               : 23.976 (24000/1001) FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 10 bits
Bits/(Pixel*Frame)                       : 0.136
Stream size                              : 6.75 GiB (95%)
Language                                 : English
Default                                  : Yes
Forced                                   : No
```

**Comment by MauriceW67** [comment]
@thexai  @Hitcher I also tested some files in Dolby Vision format.

It seems this feature only works for DV files where mediainfo shows that the HDR format is HDR10 or HDR10+ compatible:

`HDR format                               : Dolby Vision, Version 1.0, dvhe.08.06, BL+RPU, HDR10 compatible / SMPTE ST 2094 App 4, Version 1, HDR10+ Profile B compatible`

As soon as the file only shows Dolby Vision without HDR10 compatibility, it won't work:

`HDR format                               : Dolby Vision, Version 1.0, dvhe.05.06, BL+RPU`

**Comment by thexai** [comment]
Fixed not working with some Dolby Vision profiles

**Comment by thexai** [comment]
@sarbes 
I know the mapping may not be perfect but at the very least it's better than what's there now. In fact, if there are complaints it is because as it is now is incorrect, on the contrary no one has complained that there is banding in Windows that already implements this.

The most important thing here is to adjust the brightness intensity in OSD. The fact that there may be banding in the subtitles, which are generally solid white, or playback status bar I don't think is a problem.

Keep in mind that when HDR10 playback ends and is returned to the GUI, this is disabled.

**Comment by Hitcher** [comment]
> Fixed not working with some Dolby Vision profiles

Confirmed working, thanks.

**Comment by sarbes** [comment]
> @sarbes I know the mapping may not be perfect but at the very least it's better than what's there now. In fact, if there are complaints it is because as it is now is incorrect, on the contrary no one has complained that there is banding in Windows that already implements this.

Why not implement the mapping correctly in the first place? At least the part we can control is worth to get right, isn't it?

Then we only have the quantization loss.

**Comment by thexai** [comment]
An GLES shaders improvement can be made in a follow-up PR and also implement same for GL but but I'm not the one to do this and I'm not going to do it.

**Comment by sarbes** [comment]
I'm sorry, but I have to object merging the PR in this state, especially since you don't want to work on it in the future.

The current execution is wrong, the texture shaders are not the right place for adjusting the intensity. Gamma correct scaling has to be done in a post processing pass, if you want to have any resemblance of color correctness. Such an implementation would be of similar scale. And the code complexity would be less, as it would be much cleaner overall.

I want to add such a pass in order to have a HDR UI some day. I can offer you to implement a basic version which deals with the full- to limited-range conversion. From there, you could easily append your brightness mapping.

If you like, we can talk about it on our next team call.

**Comment by thexai** [comment]
> If you like, we can talk about it on our next team call.

I have a normal job of 9 hours a day and night shifts. I can't spend a lot of hours on this.

**Comment by ksooo** [comment]
Why can't we go with this PR and improve / replace this code once @sarbes (or somebody else) comes up with something real (a PR)? I tend to prefer to have something not 100% now (!) instead of getting something "perfect" in the future. What speaks against an iterative approach? Just my 2ct.

**Comment by sarbes** [comment]
Because I would need to rip out half of the code. It has no place being there (the texture shaders).

The alternative I'm proposing would be much simpler for @thexai to implement (like 6 LOC).

I don't want to discourage you. I think it is a bit unfortunate that we have not talked about it before you did this PR.

**Comment by thexai** [comment]
I am not your employee (or anything like that). Here everyone works on what they like/interest.

If you were interested in this (which you are not) you would have already implemented it.

https://forum.kodi.tv/showthread.php?tid=376360&pid=3185207#pid3185207
https://forum.kodi.tv/showthread.php?tid=376360&pid=3185128#pid3185128

**Comment by sarbes** [comment]
> > Why not implement the mapping correctly in the first place?
> 
> Can you be more specific? What is bad? and for you what is "correct mapping" (suggested changes, etc.)

See above.

> > It might also lengthen the execution time of the shaders, depending on the platform.
> 
> How a single float multiplication in shader code can slowdown execution?

The Mali-400 series can do one MAD per cycle. In the worst case, you would double the amount of cycles needed. Similar for VideoCore GPUs. 

In case of your PR, you pay the cycle cost per fragment. In case of a separate pass, you pay the cost only once.


> > Then we only have the quantization loss.
> 
> 8-bit SDR is 256 steps and in shaders is 0.0 to 1.0 range
> 
> 10-bit HDR is 1024 steps and in shaders is same 0.0 to 1.0 range
> 
> Even using only half of scale (0.5) there are 512 steps (double precise that SDR)
> 
> With default setting 40% is used 0.0 to 0.58 range
> 
> 1024 x 0.58 = 593 steps
> 
> Where is quantization loss? And why 593 steps are not sufficient to render Kodi GUI that is SDR in origin (256 steps).

You need to remember that the framebuffer is just RGBA8.

> > If you like, we can talk about it on our next team call.
> 
> I have a normal job of 9 hours a day and night shifts. I can't spend a lot of hours on this.

Me too, I can relate to that.

**Comment by thexai** [comment]
Since I have spent a few dozen hours on this PR, I hope that it will at least serve to implement this in the future. I think all the diff code is valid except the changes in shaders files (*.frag).

Then someone with more GLES shaders knowledge can take same diff and make required changes. 

PS. And probably also will fix video range conversion that is currently implemented in similar (wrong) way:

https://github.com/xbmc/xbmc/blob/213cb317db5687479e5cf1b4b5058082a6ddadb1/system/shaders/GLES/2.0/gles_shader_default.frag#L32-L35

**Comment by sarbes** [comment]
A few remarks, for documentation purposes.

The "scaling" of the non-gamma adjusted textures seems to be valid. The lack of a accompanying brightness adjustment is what threw me a bit off.

Nevertheless, the loss of color precision is severe in this implementation. Banding is very pronounced, and generally not suited for an HDR screen:
image

However, dithering the UI before quantizing produces acceptable results:
image

Both screenshots were taken from my two pass rendering implementation I'm working on, using a RGBA16f framebuffer. Quantization is simulated in the resolve shader.

To sum it up, this feature is best to be implemented in a resolve shader for quality reasons. The resolve pass has the additional benefit of being much more flexible in terms of UI quality (optional RGBA16f buffer) and rendering speed (lower resolution or frame rate on GPU video renderers).

I do find it most unfortunate how this turned out.

**Comment by thexai** [comment]
But the fact is that I can't predict the future and I can't implement a thing in "resolve shader" that doesn't exist yet 🙂

I don't mind rebase my PR over yours (but you could have done the same too).

In any case, I am using this PR on my Shield and the quality is more than acceptable (I don't see banding anywhere). I think it could have even been included in v21 RC without problems... (I don't mind waiting though).

As commented in PR, in Windows (shader) is more elaborated because is need color space conversion from BT.709 to BT.2020 and tranfer from gamma to PQ. This is not necessary in Shield because because it is done with HW.

**Comment by ChocOranger** [comment]
I don't think Sabres has a DV high Nit TV or he would be prioritising a change that stops our retinas being blinded.
Implement a 'basic' fix in this release that no-one has complained about except Sabre. 
Mark is as experimental in the Options if you're concerned. 

Sabres' fix is currently scheduled for Kodo 22...  ages away :(

@thexai if this pr doesn't make it, please do a Kodi build for each release until Sabres gets his way.

**Comment by thexai** [comment]
Reopened. At least 2 team members have expressed their opinion on Slack in favor of merging as is (and in v21):

> In a mediaplayer world, the feature you cannot offer today will drive your users away - and it is hard to get them back when feature is ready in 2026 major release only.
> 
> That being said - technically - there is one place where post processing should be done.
> 
> If someone highjacks a shader which touches every pixel anyways with a dead simple multiplication ... and can offer a working solution in 2024 - I am totally okay with it.
> 
> especially as moving on from this with the proper solution is a very simple revert of this shader part, while rest of the glue code (settings, etc.) can just stay as is.
> 
> From a user perspective: win, win
> 
> feature is there, quality will increase with the 2026 release ... settings stay the same.
> 
> That's my technical and my user-focussed opinion.

Maybe it should be put to a vote or just allow it to be merged when it receives 2 approvals.

In the forum the feedback is very positive on Shield and there is also a user who is using it on **Fire TV Stick 4K Max 2nd gen** and is happy:

https://forum.kodi.tv/showthread.php?tid=376360&pid=3186614#pid3186614

The risk is really minimal because it has already been tested on 50% of target devices.

The "mapping isn't perfect" argument doesn't hold water when what's out there now is much worse (SDR GUI at 1000 nits). For some things like subtitles there are workarounds but for others there is no possible way to mitigate it (Debug OSD, Video Info OSD...).

A PR can be required to be perfect at code level and code guileless but not in functionality, especially if the platform does not have the code base prepared for the new functionality (and who opens the PR is not the main platform developer or maintainer).

On the other hand, this is a missing functionality in Android since it already exists in Windows and is functional since v19... this PR only tries to short the big difference.

**Comment by thexai** [comment]
For my part, I thank sarbes for taking this seriously and starting to quickly implement this functionality for v22. It will undoubtedly be a great improvement, although also a much more invasive change: it affects other platforms besides Android and must be thoroughly tested in case there are performance regressions. On the contrary, in this PR there are enough guarantees that not affect other platforms.

**Comment by thexai** [comment]
Rebased only, no code changes.

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererMediaCodecSurface.cpp [hunk 1]
```diff
@@ -63,6 +63,18 @@ bool CRendererMediaCodecSurface::Configure(const VideoPicture &picture, float fp
...
+  if (picture.color_transfer == AVCOL_TRC_SMPTE2084)
+  {
+    if (CServiceBroker::GetWinSystem()->IsHDRDisplay())
+      CServiceBroker::GetWinSystem()->GetGfxContext().SetTransferPQ(true);
+  }
+  else if (picture.hdrType == StreamHdrType::HDR_TYPE_DOLBYVISION)
+  {
+    if (CServiceBroker::GetWinSystem()->GetDisplayHDRCapabilities().SupportsDolbyVision())
+      CServiceBroker::GetWinSystem()->GetGfxContext().SetTransferPQ(true);
+  }
```
[the remark this answers is not in the document]
Comment (CrystalP) [reply xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererMediaCodecSurface.cpp:76]: Also is there a similar code path to modify for software decoding?

**Comment by CrystalP** [comment]
I think the luminance mapping is good enough because the multiplier is calculated using a linear equation in CWinSystemAndroid::GetGuiSdrPeakLuminance() and the multiplied data is in PQ, so it's perceptually linear.

Unlike Windows where the multiplication is in linear space so the curve of the multiplier is exponential for a similar effect.

Is the fact that the primaries are converted by Android reliable? Even if not I guess not being blinded is better that inaccurate colors. Doing custom conversion to be certain would obviously be much more expensive for the GPU (similar to the Windows shader code).

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererMediaCodecSurface.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (thexai) [reply xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererMediaCodecSurface.cpp:76]: In Android there are not much options: Android devices are not powerful enough to do software HDR video decoding / rendering. If no HDR or DV display available then devices _may_ do HDR to SDR mapping (HW based) but this is already taken into account as then `SetTransferPQ(true)` is not reached.

GUI Setting (old name) "Use HDR display capabilities" is half broken in Android because when MediaCodecSurface is used, output is always "HDR passthrough" if system has HDR display (even with setting disabled).

Intentionally only is handled RenderMediCodecSurface code path because SWR path anyway is not usable for HDR on Android devices.

**Comment by thexai** [comment]
> Is the fact that the primaries are converted by Android reliable? Even if not I guess not being blinded is better that inaccurate colors.

Colors are totally correct (at least on Shield).

**Comment by sarbes** [comment]
> But the fact is that I can't predict the future and I can't implement a thing in "resolve shader" that doesn't exist yet 🙂

I'm pretty vocal about my ideas and changes in our Slack, I would say. I also have quite a bunch of them written down as a project (https://github.com/orgs/xbmc/projects/5). But unfortunately, it seems that I'm mostly alone in both conversations. Aside from my current PRs, I do have quite a bunch of uncommitted code, which is waiting of the approval of the former. I'm always available to talk about changes falling in the GUI/GL/GLES domain.


> I don't think Sabres has a DV high Nit TV or he would be prioritising a change that stops our retinas being blinded. 

Those are assumptions on your side. My assumption of you is that you don't grasp the base of my objection.

> Implement a 'basic' fix in this release that no-one has complained about except Sabre. Mark is as experimental in the Options if you're concerned.

Make no mistake, this is not a fix but a workaround for broken devices/software. 

My "complaint" was never about improving the user experience of this release, but the execution. And I don't think that a lot of people are qualified to have remarks about the latter. So naturally, the pool of such people (thus the amount of "complaints") is small compared to our large and rather vocal user base.

> Sabres' fix is currently scheduled for Kodo 22... ages away :(

It could be incorporated into 21. But we don't have many reviewers and it might be considered to big of a change for the beta. Personally, I'd be willing to risk it.


> Reopened. At least 2 team members have expressed their opinion on Slack in favor of merging as is (and in v21):
> ...
> Maybe it should be put to a vote or just allow it to be merged when it receives 2 approvals.

You have my vote as well, if you would consider working on it once the post processing pass gets merged. I don't have the setup and the time needed to qualify changes. My main objection was (is) mainly about the idea that I would need to take ownership about something I can't test.

**Comment by sarbes** [comment]
> > Is the fact that the primaries are converted by Android reliable? Even if not I guess not being blinded is better that inaccurate colors.
> 
> Colors are totally correct (at least on Shield).

While I don't have a suitable setup to test, I would expect it to be right (for most systems).

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererMediaCodecSurface.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (CrystalP) [reply xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererMediaCodecSurface.cpp:76]: Are there PVR type of situations where a stream could change back and forth between HDR and SDR? If so, need an `else SetTransferPQ(false);` or similar.

I was wondering about dovi fallback to hdr10 but that seems covered already, right ?(streams that have HDR10 compatibility)

HLG streams likely need the luminance adjustment too.

[the remark this answers is not in the document]
Comment (thexai) [reply xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererMediaCodecSurface.cpp:76]: > Are there PVR type of situations where a stream could change back and forth between HDR and SDR? If so, need an else SetTransferPQ(false); or similar.

`CRendererMediaCodecSurface::Configure` only is called one time when opening a stream.

> HLG streams likely need the luminance adjustment too. 

Not in Android. This was already tested. In Windows is need because not supports HLG output and is converted to HDR10 (when display supports HDR). In Android, GUI has correct bright, while playing HLG streams before and after this PR.

**Comment by thexai** [comment]
> You have my vote as well, if you would consider working on it once the post processing pass gets merged. I don't have the setup and the time needed to qualify changes. My main objection was (is) mainly about the idea that I would need to take ownership about something I can't test.

I will work on it keeping in mind that GL/GLES is not my area and I also have limited time. I will try to do the best I can.

**Comment by sarbes** [comment]
> I will work on it keeping in mind that GL/GLES is not my area and I also have limited time. I will try to do the best I can.

That's a statement I can perfectly live with.

I'll approve the PR from the the GLES side of things. It's probably best to get the green light from an Android guy and the release folks before merging.

**Review by sarbes — APPROVED** [review]
Approving the GLES changes.

**Comment by hugbug** [comment]
I noticed the brightness adjustment is activated depending on content instead of display mode. That works as long as the Fire TV option "Dynamic range settings" is set to "Adaptive".

With "Dynamic range settings" set to "Always HDR", Kodi UI is very bright until a HDR video is started. After stopping the video and returning to Kodi UI the full brightness is back.

**Comment by sarbes** [comment]
That's the idea. The adjustment is just for playback.

Kodi has no HDR mode (or similar) for the UI.

**Comment by Hitcher** [comment]
Yeah, you need to turn off 'Always HDR'.

**Review by fritsch — APPROVED** [review]
My reasoning was already cited (anonymously ;-))

**Comment by MauriceW67** [comment]
@thexai Am I correct in assuming that the option does not work for playing DV streams through Input Stream Adaptive, such as with the Disney+ addon?

**Comment by thexai** [comment]
No idea, but should work if video is currently played as Dolby Vision on TV: i.e. DV metadata is passed to MediaCodecSurface and DV capable display is identified inside Kodi settings information.

**Comment by MauriceW67** [comment]
> No idea, but should work if video is currently played as Dolby Vision on TV: i.e. DV metadata is passed to MediaCodecSurface and DV capable display is identified inside Kodi settings information.

Strange. I played an episode of Secret Invasion through the Disney+ addon and my TV definitely switched to DV mode. But the OSD and subtitles stayed on the SDR brightness level.

Playing local DV content in MP4 or MKV format from my NAS works fine though.

**Comment by Hitcher** [comment]
> Strange. I played an episode of Secret Invasion through the Disney+ addon and my TV definitely switched to DV mode. But the OSD and subtitles stayed on the SDR brightness level.
> 
> Playing local DV content in MP4 or MKV format from my NAS works fine though.

I can confirm the same behaviour playing a DV video in the D+ addon.

**Comment by thexai** [comment]
Add-on thing will be investigated later. May be necessary modify something in `InputStreamAddon.cpp` and require add-ons API version bump anyway (for v22 only).

The first step to continue working in this is merge this PR now...

**Comment by ilikenwf** [comment]
@thexai how hard would it be to add this to Linux as well? I use HDR with Kodi on Linux, my own builds, and can see how this would be a nice feature so as not to burn my eyes.

**Comment by josh4trunks** [comment]
> @thexai how hard would it be to add this to Linux as well? I use HDR with Kodi on Linux, my own builds, and can see how this would be a nice feature so as not to burn my eyes.

I would appreciate this coming to Linux on x86_64 as well!
LibreELEC user here, and ever since getting an LG OLED a few months ago I realized GUI/SRT subtitles were an issue with HDR/DV content.
